Clear card present payment onboarding cache after logging out

staskus opened the pull request:
When we onboard card present payments, log out, and log in again, an outdated but invalid card present payment cache is still available.

<!-- Remember about a good descriptive title. -->

Closes: #13842
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

When testing POS, we noticed that connection to the reader fails with 2 payment providers and using POS after logging out and logging in.

It happens because `CardPresentPaymentOnboardingStateCache` is not cleared after logging out. It's only cleared after switching stores.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Log into an account with 2 payment providers (Stripe, WooPayments)
2. Go to menu
3. Confirm no POS option appears
4. Open Payment menu
5. Finalize onboarding by selecting WooPayments as the main provider
6. Confirm POS option appears
7. Settings -> Log out
8. Log into the same account
9. Confirm no POS option appears, and onboarding has to be again finalized in Payments menu

## Testing Information

Tested on iOS 17:
- A site with two payment providers
- A site with one payment provider
- Switching sites
- Killing & relaunching the app to confirm that the cache remains
- Went through the cases in the code when deauthenticate() is called, to confirm that this is the main method that is called in various unhappy path corner-cases when users are logged out

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/98406aa9-b0d8-4404-8960-8d71c2610c87

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Clear card present payment onboarding state cache when deauthenticating

When we onboard card present payments, log out, and log in again, an outdated but invalid card present payment cache is still available.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald reviewed (commented):
Code looks good and tests well.

I checked the steps listed on an iPad running iOS 16.

I also checked TTP and auto-reconnection on an iPhone running iOS 17

What manual testing did you do, for the purposes of the checklist?

staskus committed:
Use CardPresentConfigurationLoader within CardPresentPluginsDataProvider to always use latest configuration

Use CardPresentConfigurationLoader within CardPresentPluginsDataProvider to always use latest configuration when determining if plugin version is supported. Initially, configuration can be loaded with SiteAddress settings which later change.

staskus commented:
@joshheald thanks, I listed them in `Testing Information` section.

One other thing I noticed is that after I log in, this method returns `false` until I relaunch the app or enter `Payments` menu.
It happens because `configuration.supportedPluginVersions` is empty.  `CardPresentPluginsDataProvider` is initialized with static configuration, which remains outdated. Initially, when the configuration is created, the country code is `unknown`.

```
    private func isPluginVersionSupported(plugin: Yosemite.SystemPlugin,
                                          paymentPlugin: CardPresentPaymentsPlugin) -> Bool {
        guard let pluginSupport = configuration.supportedPluginVersions.first(where: { support in
            support.plugin == paymentPlugin
        }) else {
            return false
        }
        return VersionHelpers.isVersionSupported(version: plugin.version,
                                                 minimumRequired: pluginSupport.minimumVersion)
    }
 ```

The solution would be to use `configurationLoader` within `CardPresentPluginsDataProvider` to make sure the latest one is used.

I included the change in this PR https://github.com/woocommerce/woocommerce-ios/pull/13897/commits/4c7b3fe27227d1dcf4bb00748fbe69c231e8004c. I don't see dangers in using configurationLoader rather than configuration. However, let me know if you would like this to be made in a separate PR with more testing considerations.

joshheald reviewed (approved):
> I included the change in this PR [4c7b3fe](https://github.com/woocommerce/woocommerce-ios/commit/4c7b3fe27227d1dcf4bb00748fbe69c231e8004c). I don't see dangers in using configurationLoader rather than configuration. However, let me know if you would like this to be made in a separate PR with more testing considerations.

Good catch. I agree, using the loader to avoid the race condition seems low risk to me.

I think all that's left is to mention these changes in the release notes (and fixing a lint error.)

staskus committed:
Fix Swiftlint issue

staskus committed:
Update RELEASE-NOTES.txt

joshheald commented on WooCommerce/Classes/Tools/CardPresentPluginsDataProvider.swift:
```diff
@@ -30,16 +30,18 @@ protocol CardPresentPluginsDataProviderProtocol {
-    private let configuration: CardPresentPaymentsConfiguration
+    private let configurationLoader: CardPresentConfigurationLoader
+
+    static let shared = CardPresentPluginsDataProvider(configurationLoader: CardPresentConfigurationLoader(stores: ServiceLocator.stores))
```
@staskus Is this singleton absolutely required? I can see it's only used in the receipt parameters provider and can't find any issues with it in my beta testing immediately... but I wonder how well it survives store changes between different countries... at least when generating receipts?

staskus replied:
Hey, good question. `CardPresentPluginsDataProvider` only holds `CardPresentConfigurationLoader` which in turn doesn't keep any state and returns a configuration. However, since it's not required we could do without this singleton.

joshheald replied:
If we can do without it, I think we should. I don't think it causes any issues now, but it's the kind of thing which might in future as the code is changed. 

The StripeCardReaderService is already a singleton, so having more than one interacting with each other could get quite messy. Would you mind changing it?

staskus replied:
Of course. I changes it here:

https://github.com/woocommerce/woocommerce-ios/pull/13982

I also retested `PaymentReceiptEmailParameterDeterminer` on `trunk` to make sure there are no issues even without a fix.
